xarray/core/variable.py

```diff
@@ -637,17 +637,22 @@ def __setitem__(self, key, value):
         """
         dims, index_tuple, new_order = self._broadcast_indexes(key)
 
-        if isinstance(value, Variable):
-            value = value.set_dims(dims).data
-
-        if new_order:
-            value = duck_array_ops.asarray(value)
+        if not isinstance(value, Variable):
+            value = as_compatible_data(value)
             if value.ndim > len(dims):
                 raise ValueError(
```

Comment:
Do we still need this special case error message now that we call set_dims below?

Reply:
I think `value = Variable(dims[-value.ndim:], value)` fails if `value.ndim > len(dims)`.

Reply:
I think `value = Variable(dims[-value.ndim:], value)` fails if `value.ndim > len(dims)`.